### PR TITLE
[UIMA-6199] Unindexed document annotation added back to index during deserialization

### PR DESCRIPTION
**Pull request by reckart**
JIRA Issue: https://issues.apache.org/jira/browse/UIMA-6199

- Added a unit test demonstrating the issue (currently failing of course)

### CODE REVIEW

**Comment by mischor** [comment]
committed a fix.  let's see if the tests pass.

**Comment by mischor** [comment]
@reckart do you want to review this?

File: uimaj-core/src/main/java/org/apache/uima/cas/impl/BinaryCasSerDes.java [hunk 1]
```diff
@@ -973,6 +974,8 @@ private SerialFormat binaryDeserialization(Header h) {
         msg = e.toString();
       }
       throw new CASRuntimeException(CASRuntimeException.BLOB_DESERIALIZATION, msg);
+    } finally {
+        baseCas.forAllViews(view -> view.set_deserialized_doc_annot_not_indexed(null));
```
Comment (reckart) [inline uimaj-core/src/main/java/org/apache/uima/cas/impl/BinaryCasSerDes.java]: A comment could be nice here explaining why this is done.

Comment (reckart) [reply uimaj-core/src/main/java/org/apache/uima/cas/impl/BinaryCasSerDes.java]: Why does the setter not use the typical Java naming conventions?

File: uimaj-core/src/main/java/org/apache/uima/cas/impl/BinaryCasSerDes.java [hunk 2]
```diff
@@ -1697,12 +1700,23 @@ private void createFSsFromHeaps(boolean isDelta, int startPos, CommonSerDesSeque
...
           Sofa sofa = getSofaFromAnnotBase(heapIndex, stringHeap, addr2fs, csds);  // creates sofa if needed and exists (forward ref case)
           view = (sofa == null) ? baseCas.getInitialView() : baseCas.getView(sofa);
           if (type == tsi.docType) {
-            fs = view.getDocumentAnnotation();  // creates the document annotation if it doesn't exist
-            view.removeFromCorruptableIndexAnyView(fs, view.getAddbackSingle());
+            Annotation documentAnnotationPrevious = view.getDocumentAnnotationNoCreate();
+            if (documentAnnotationPrevious == null) {
+              // document annotation not present
+              fs = view.createDocumentAnnotationNoRemoveNoIndex(0); // create but don't index
+              view.set_deserialized_doc_annot_not_indexed((Annotation)fs);  // for use by other code that sets length, if this is not indexed
```
Comment (reckart) [inline uimaj-core/src/main/java/org/apache/uima/cas/impl/BinaryCasSerDes.java]: length of what?

File: uimaj-core/src/main/java/org/apache/uima/cas/impl/BinaryCasSerDes.java [hunk 3]
```diff
@@ -1697,12 +1700,23 @@ private void createFSsFromHeaps(boolean isDelta, int startPos, CommonSerDesSeque
...
+              // documentAnnotationPreviouslyIndex == false, preset above
+            } else {
+                fs = documentAnnotationPrevious;
```
Comment (reckart) [inline uimaj-core/src/main/java/org/apache/uima/cas/impl/BinaryCasSerDes.java]: I believe there is no code which prevents the user from manually instantiating multiple document annotation instances and adding these to the indexes. Would the code support that or mess up?

File: uimaj-core/src/main/java/org/apache/uima/cas/impl/CASImpl.java [hunk 4]
```diff
@@ -1117,6 +1117,10 @@ public boolean isCasLocked() {
    */
   private volatile FSIterator<Annotation> docAnnotIter = null;
   
+  // UIMA-6199 provides access to non-indexed doc annot 
+  //   to allow sofa setting to set the "length"
+  private volatile Annotation deserialized_doc_annot_not_indexed = null;
```
Comment (reckart) [inline uimaj-core/src/main/java/org/apache/uima/cas/impl/CASImpl.java]: Length of what?

File: uimaj-core/src/main/java/org/apache/uima/cas/impl/BinaryCasSerDes.java (hunk 1 above)
Comment (mischor) [reply uimaj-core/src/main/java/org/apache/uima/cas/impl/BinaryCasSerDes.java]: I'll add a comment.  re: naming conventions: I'm trying out "snake" case as an alternative to "camel" case.  I seen reports that for non-native English readers, it's a bit easier to follow.

Comment (reckart) [reply uimaj-core/src/main/java/org/apache/uima/cas/impl/BinaryCasSerDes.java]: Snake-case is typical for Python - but in Java code, I'd rather opt to sticking with the traditional Java conventions.

File: uimaj-core/src/main/java/org/apache/uima/cas/impl/BinaryCasSerDes.java (hunk 2 above)
Comment (mischor) [reply uimaj-core/src/main/java/org/apache/uima/cas/impl/BinaryCasSerDes.java]: The Document Annotation is a subtype of "Annotation" and has a begin and end.  The begin is normally defaulted to 0.  When the sofa data is a "string", the "end" is updated with the length of the sofa data string.  (although I cannot find this in the docs, it's what the framework has been doing forever...)

File: uimaj-core/src/main/java/org/apache/uima/cas/impl/BinaryCasSerDes.java (hunk 3 above)
Comment (mischor) [reply uimaj-core/src/main/java/org/apache/uima/cas/impl/BinaryCasSerDes.java]: I think the code mostly ignores this.  There's a method in the cas to get the DocumentAnnotation, and it runs an annotation iterator over that type in the particular view, and uses the 1st one.   I'm pretty sure no one has thought about that case, though.

File: uimaj-core/src/main/java/org/apache/uima/cas/impl/CASImpl.java (hunk 4 above)
Comment (mischor) [reply uimaj-core/src/main/java/org/apache/uima/cas/impl/CASImpl.java]: I'll add a comment.  This is the length of the sofa data when it's a string.  (Have to look to see if it's set at all for the other forms of sofa data...)

**Comment by mischor** [comment]
A general observation: I see now in the code that the framework does attempt to keep just 1 DocumentAnnotation Instance.  You can see this in the CASImpl internal createDocumentAnnotation method, where it goes and "removes from the indexes" any others that might be there, first.

**Comment by mischor** [comment]
If no objections, I'll commit to master...

**Pull request merged**